[project-base] ajaxMoreLoader is generalized (#579)
shopsys/shopsys: docs/cookbook/adding-ajax-load-more-button-into-pagination.md

# Adding Ajax Load More Button into Pagination

> This tutorial will guide us via steps needed for implementing page loader that will make calls to `Brand` controller. 

Reviewer: I would put it this way: "In this cookbook, we will add a paginated brand list including ajax "load more" button to a product list page. After finishing the guide, you will know how to use multiple paginations on one page."

{% endblock %}
```

```twig
{# ShopBundle/Resources/views/Front/Content/Brand/ajaxList.html.twig #}
{% import '@ShopsysShop/Front/Inline/Paginator/paginator.html.twig' as paginator %}
{% set entityName = 'brands'|trans %}
{% set pageQueryParameter = 'brandPage' %}

<div>
    <div class="js-list-with-paginator">
        {{ paginator.paginatorNavigation(paginationResult, entityName, pageQueryParameter) }}
        <ul class='list-images js-list'>
            {% for brand in paginationResult.results %}
                <li class="list-images__item">
                    <a href="{{ url('front_brand_detail', { id: brand.id }) }}" class="list-images__item__block list-images__item__block--with-label">
                        {{ image(brand, { alt: brand.name }) }}
                        <span>{{ brand.name }}</span>
                    </a>
                </li>
            {% endfor %}
        </ul>
        <div class="text-center margin-bottom-20">
            {{ paginator.loadMoreButton(paginationResult, url('front_brand_list'), pageQueryParameter) }}
        </div>
        {{ paginator.paginatorNavigation(paginationResult, entityName, pageQueryParameter) }}
    </div>
</div>
```

> After that we will modify `listAction` method in `BrandController` so `Brand` list page will be paginated and we will be able to integrate it into another list page that has other paginated items.  

Reviewer: I am thinking - by modifying the `listAction` you will modify `brands-list` page as well - it will be paginated as well as a side effect :slightly_smiling_face: It did not get it from the first sight. So I would either mention it here or create a new action for that purpose

 * @param mixed $drawLayout
 */
public function listAction(Request $request, $drawLayout = true)
{
    if ($request->isXmlHttpRequest() || $drawLayout === false) {
        $template = '@ShopsysShop/Front/Content/Brand/ajaxList.html.twig';
    } else {
        $template = '@ShopsysShop/Front/Content/Brand/list.html.twig';
    }

    $requestPage = $request->get(self::PAGE_QUERY_PARAMETER);
    $page = $requestPage === null ? 1 : (int)$requestPage;

    return $this->render($template, [
        'paginationResult' => $this->brandFacade->getPaginatedResult($page, self::ITEMS_PER_PAGE),
    ]);
}
```

## Integration of Paginated Brand List

Now we have working implementation of paginated `Brand` list page that can be loaded also from asynchronous calls.
We can try to integrate it into another `Product` list page that is also paginated with page query parameter `page`.
Only thing we need to do is to modify template for `Product` page.  
We will add twig code into `main_content` block.

```twig
{# ShopBundle/Resources/views/Front/Content/Product/list.html.twig #}
{{ render(controller('ShopsysShopBundle:Front/Brand:list', { drawLayout: false })) }}
```

## Conclusion

> Customer can see 2 paginated lists with buttons for loading items from next pages on each `Product` list page. Since there are unique page query parameters, paginated lists can have displayed different page indexes after broser is loaded with these page query parameters.

Reviewer: typo "broser"

# Adding Ajax Load More Button into Pagination

In this cookbook, we will add a paginated brand list including ajax "load more" button to a product list page. After finishing the guide, you will know how to use multiple paginations on one page.  

> *Note: After this change you will have paginated also `/brands-list/` route.*

Reviewer: ``
```suggestion
*Note: After this change you will have paginated also `/brands-list/` page (`front_brand_list` route).*
```

## Implementation of Brand Pagination

> First we need to extend BrandRepository by creating `BrandRepository.php` in `/src/Shopsys/ShopBundle/Model/Product/Brand` and we add `getPaginationResult` method.

Reviewer: ```suggestion
First we need to extend `Shopsys\FrameworkBundle\Model\Product\Brand\BrandRepository` by creating `BrandRepository.php` in `/src/Shopsys/ShopBundle/Model/Product/Brand` and we add `getPaginationResult` method.
```

```php
/**

>  * {@inheritDoc}

Reviewer: this is not inherited method

 */
public function getPaginationResult(
    $page,
    $limit
) {
    $queryBuilder = $this->getBrandRepository()->createQueryBuilder('b');
    $queryBuilder->orderBy('b.name', 'asc');

    /** @var \Shopsys\FrameworkBundle\Component\Paginator\QueryPaginator $queryPaginator */
    $queryPaginator = new QueryPaginator($queryBuilder);

    return $queryPaginator->getResult($page, $limit);
}
```

Then we will extend `BrandFacade` and add `getPaginatedResult` method.
```php
/**

>  * {@inheritDoc}

Reviewer: this is not an inherited method as well, right?

> Then we will extend `BrandFacade` and add `getPaginatedResult` method.

Reviewer: ```suggestion
Then we will extend `Shopsys\FrameworkBundle\Model\Product\Brand\BrandFacade` and add `getPaginatedResult` method.
```

> /**

Reviewer: ```suggestion
// use Shopsys\FrameworkBundle\Model\Product\Brand\BrandFacade as BaseBrandFacade
// Shopsys\ShopBundle\Model\Product\Brand\BrandFacade extends BaseBrandFacade

/**
```

 */
public function getPaginatedResult(
    $page,
    $limit
) {
    $paginationResult = $this->brandRepository->getPaginationResult(
        $page,
        $limit
    );

    return new PaginationResult(
        $paginationResult->getPage(),
        $paginationResult->getPageSize(),
        $paginationResult->getTotalCount(),
        $paginationResult->getResults()
    );
}
```

> After we have extended `BrandRepository` and `BrandFacade` we will create aliases for them in `services.yml`.

Reviewer: ```suggestion
After we have extended `BrandRepository` and `BrandFacade` we need to set them to be used instead of the framework classes in our application. This is done via configuration in `services.yml`.
```

{% set pageQueryParameter = 'brandPage' %}

<div>
    <div class="js-list-with-paginator">
        {{ paginator.paginatorNavigation(paginationResult, entityName, pageQueryParameter) }}
        <ul class='list-images js-list'>
            {% for brand in paginationResult.results %}
                <li class="list-images__item">
                    <a href="{{ url('front_brand_detail', { id: brand.id }) }}" class="list-images__item__block list-images__item__block--with-label">
                        {{ image(brand, { alt: brand.name }) }}
                        <span>{{ brand.name }}</span>
                    </a>
                </li>
            {% endfor %}
        </ul>
        <div class="text-center margin-bottom-20">
            {{ paginator.loadMoreButton(paginationResult, url('front_brand_list'), pageQueryParameter) }}
        </div>
        {{ paginator.paginatorNavigation(paginationResult, entityName, pageQueryParameter) }}
    </div>
</div>
```

After that we will modify `listAction` method in `BrandController` so `Brand` list page will be paginated and we will be able to integrate it into another list page that has other paginated items.  

> We will implement also constants for page query parameter `const PAGE_QUERY_PARAMETER = 'brandPage'` and for the count of items on one page `const ITEMS_PER_PAGE = 5;`

Reviewer: ```suggestion
We will implement also constants for page query parameter `const PAGE_QUERY_PARAMETER = 'brandPage'` and for the count of items on one page `const ITEMS_PER_PAGE = 5;`.

To be able to determine whether the request for brand list is called from the template, we need to add dependecy on `Symfony\Component\HttpFoundation\RequestStack` into our `BrandController`.
```

```yaml
    Shopsys\FrameworkBundle\Model\Product\Brand\BrandRepository: '@Shopsys\ShopBundle\Model\Product\Brand\BrandRepository'

    Shopsys\FrameworkBundle\Model\Product\Brand\BrandFacade: '@Shopsys\ShopBundle\Model\Product\Brand\BrandFacade'
```

Next we will modify the brand list twig template and create twig template for rendering paging controls and paginated items via ajax.  
We will use `paginator.loadMoreButton(paginationResult, url('front_brand_list'), pageQueryParameter)` twig component that will asynchronously load next page when user clicks on its button. We will also define `pageQueryParameter` variable so it will have unique name and it will not interfere with other paging component on the same page.

```twig
{# ShopBundle/Resources/views/Front/Content/Brand/list.html.twig #}
{% extends '@ShopsysShop/Front/Layout/layoutWithPanel.html.twig' %}

{% block title %}
    {{ 'Brand overview'|trans }}
{% endblock %}

{% block main_content %}
    <div>
        <h1>{{ 'Brand overview'|trans }}</h1>

>         {% include '@ShopsysShop/Front/Content/Brand/ajaxList.html.twig' with {paginationResult: paginationResult} %}

Reviewer: ```suggestion
        {% include '@ShopsysShop/Front/Content/Brand/ajaxList.html.twig' with {paginationResult: paginationResult} %}
        {# ... #}
```

>         {% include '@ShopsysShop/Front/Content/Brand/ajaxList.html.twig' with {paginationResult: paginationResult} %}

Reviewer: ```suggestion
        {% include '@ShopsysShop/Front/Content/Brand/ajaxList.html.twig' with {paginationResult: paginationResult} %}
        {# ... #}
```

                        {{ image(brand, { alt: brand.name }) }}
                        <span>{{ brand.name }}</span>
                    </a>
                </li>
            {% endfor %}
        </ul>
        <div class="text-center margin-bottom-20">
            {{ paginator.loadMoreButton(paginationResult, url('front_brand_list'), pageQueryParameter) }}
        </div>
        {{ paginator.paginatorNavigation(paginationResult, entityName, pageQueryParameter) }}
    </div>
</div>
```

After that we will modify `listAction` method in `BrandController` so `Brand` list page will be paginated and we will be able to integrate it into another list page that has other paginated items.  
We will implement also constants for page query parameter `const PAGE_QUERY_PARAMETER = 'brandPage'` and for the count of items on one page `const ITEMS_PER_PAGE = 5;`.

To be able to determine whether the request for brand list is called from the template, we need to add dependecy on `Symfony\Component\HttpFoundation\RequestStack` into our `BrandController`.

```php

> /**

Reviewer: ```suggestion

const PAGE_QUERY_PARAMETER = 'brandPage';
const ITEMS_PER_PAGE = 5;

/**
```

 * @var \Symfony\Component\HttpFoundation\RequestStack
 */
private $requestStack;

/**
 * @param \Shopsys\FrameworkBundle\Model\Product\Brand\BrandFacade $brandFacade
 * @param \Symfony\Component\HttpFoundation\RequestStack $requestStack
 */
public function __construct(
    BrandFacade $brandFacade,
    RequestStack $requestStack
) {
    $this->brandFacade = $brandFacade;
    $this->requestStack = $requestStack;
}

> ...

Reviewer: ```suggestion

```

{% endblock %}
```

> There are two important css classes that must be used.

Reviewer: awesome :heart:
